src/lib/int.ts

```diff
@@ -301,119 +299,183 @@ export class UInt32 extends CircuitValue {
+    // constant case - just return unchecked value
+    if (x.isConstant()) return getUnchecked();
+    // variable case - create a new checked witness and prove consistency with original field
+    let xInt = Circuit.witness(Int64, getUnchecked);
+    xInt.toField().assertEquals(x); // sign(x) * |x| === x
+    return xInt;
   }
 
-  static sizeInFields(): number {
-    return 1;
+  neg() {
+    // doesn't need further check if `this` is valid
+    return new Int64(this.magnitude, this.sgn.neg());
   }
 
-  neg(): Int64 {
-    return new Int64(this.value.neg());
+  add(y: Int64 | number | string | bigint | UInt64 | UInt32) {
+    let y_ = argToInt64(y);
+    return Int64.fromField(this.toField().add(y_.toField()));
   }
-
-  static add(x: Int64, y: Int64 | UInt32 | UInt64) {
-    return new Int64(x.value.add(y.value));
+  sub(y: Int64 | number | string | bigint | UInt64 | UInt32) {
+    let y_ = argToInt64(y);
+    return Int64.fromField(this.toField().sub(y_.toField()));
   }
-
-  static sub(x: Int64, y: Int64 | UInt32 | UInt64) {
-    return new Int64(x.value.sub(y.value));
+  mul(y: Int64 | number | string | bigint | UInt64 | UInt32) {
+    let y_ = argToInt64(y);
+    return Int64.fromField(this.toField().mul(y_.toField()));
   }
-
-  add(y: Int64 | UInt32 | UInt64) {
-    return Int64.add(this, y);
+  div(y: Int64 | number | string | bigint | UInt64 | UInt32) {
+    let y_ = argToInt64(y);
+    let [q] = this.magnitude.divMod(y_.magnitude);
+    let sign = this.sgn.mul(y_.sgn);
+    return new Int64(q, sign);
   }
-
-  sub(y: Int64 | UInt32 | UInt64) {
-    return Int64.add(this, y);
+  mod(y: UInt64 | number | string) {
+    let y_ = new UInt64(argToField('Int64.mod', y));
+    let [, r] = this.magnitude.divMod(y_);
+    r = Circuit.if(this.isPositive(), r, y_.sub(r));
+    return new Int64(r, Field.one);
   }
 
-  repr(): { magnitude: Field; isPos: Sgn } {
-    throw 'repr';
+  equals(y: Int64 | number | string | bigint | UInt64 | UInt32) {
+    let y_ = argToInt64(y);
+    return this.toField().equals(y_.toField());
   }
-
-  static toFields(x: Int64): Field[] {
-    return [x.value];
+  assertEquals(y: Int64 | number | string | bigint | UInt64 | UInt32) {
+    let y_ = argToInt64(y);
+    this.toField().assertEquals(y_.toField());
   }
-
-  static ofFields(xs: Field[]) {
-    return new Int64(xs[0]);
+  isPositive() {
+    return this.sgn.equals(Field.one);
   }
+}
 
-  toFields(): Field[] {
-    return Int64.toFields(this);
-  }
+function argToInt64(
+  x: number | string | bigint | UInt64 | UInt32 | Int64
+): Int64 {
+  if (typeof x === 'number') return Int64.fromNumber(x);
+  if (typeof x === 'string') return Int64.fromString(x);
+  if (typeof x === 'bigint') return Int64.fromBigInt(x);
+  if (x instanceof UInt64 || x instanceof UInt32) return Int64.fromUnsigned(x);
+  return x;
+}
 
-  sizeInFields(): number {
-    return Int64.sizeInFields();
+function argToField(
+  name: string,
+  x: { value: Field } | number | string
```

Comment:
TODO: add range checks (not constraints, just checks) on constant UInt64 / UInt32 / Int64 as you create them